dali/pipeline/operators/decoder/nvjpeg_decoder_gpu.h

```diff
@@ -0,0 +1,161 @@
+    NVJPEG_CALL(nvjpegDecoderCreate(handle_, NVJPEG_BACKEND_GPU_HYBRID, &decoder_hybrid_));
+
+    NVJPEG_CALL(nvjpegBufferDeviceCreate(handle_, nullptr, &device_buffer_));
+
+    // Do we really need both in both stages ops?
+    size_t device_memory_padding = spec.GetArgument<Index>("device_memory_padding");
+    size_t host_memory_padding = spec.GetArgument<Index>("host_memory_padding");
+    NVJPEG_CALL(nvjpegSetDeviceMemoryPadding(device_memory_padding, handle_));
+    NVJPEG_CALL(nvjpegSetPinnedMemoryPadding(host_memory_padding, handle_));
+  }
+
+  ~nvJPEGDecoderGPU() noexcept(false) {
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_host_));
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_hybrid_));
+    NVJPEG_CALL(nvjpegDestroy(handle_));
+  }
+
+  using dali::OperatorBase::Run;
+  void Run(MixedWorkspace *ws) override {
+    std::vector<Dims> output_shape(batch_size_);
+    // Creating output shape and setting the order of images so the largest are processed first
+    // (for load balancing)
+    std::vector<std::pair<size_t, size_t>> image_order(batch_size_);
+    for (int i = 0; i < batch_size_; i++) {
+      const auto& info_tensor = ws->Input<CPUBackend>(0, i);
+      const auto& state_tensor = ws->Input<CPUBackend>(1, i);
+      const EncodedImageInfo* info;
+      std::tie(info, std::ignore) = GetInfoState(info_tensor, state_tensor);
+      int c = GetOutputPitch(output_image_type_);
+      output_shape[i] = Dims({info->heights[0], info->widths[0], c});
+      image_order[i] = std::make_pair(Volume(output_shape[i]), i);
+    }
+    std::sort(image_order.begin(), image_order.end(),
+              std::greater<std::pair<size_t, size_t>>());
+
+    auto& output = ws->Output<GPUBackend>(0);
+    output.Resize(output_shape);
+    TypeInfo type = TypeInfo::Create<uint8_t>();
+    output.set_type(type);
+
+    for (int idx = 0; idx < batch_size_; idx++) {
+      const int i = image_order[idx].second;
+
+      const auto& info_tensor = ws->Input<CPUBackend>(0, i);
+      const auto& state_tensor = ws->Input<CPUBackend>(1, i);
+      const EncodedImageInfo* info;
+      const StateNvJPEG* nvjpeg_state;
+      std::tie(info, nvjpeg_state) = GetInfoState(info_tensor, state_tensor);
+
+      auto *output_data = output.mutable_tensor<uint8_t>(i);
+      if (info->nvjpeg_support) {
+        nvjpegImage_t nvjpeg_image;
+        nvjpeg_image.channel[0] = output_data;
+        nvjpeg_image.pitch[0] = GetOutputPitch(output_image_type_) * info->widths[0];
+
+        nvjpegJpegState_t state = GetNvjpegState(*nvjpeg_state);
+
+        NVJPEG_CALL(nvjpegStateAttachDeviceBuffer(state,
+                              device_buffer_));
+
+        nvjpegJpegDecoder_t decoder = GetDecoder(nvjpeg_state->nvjpeg_backend);
+        NVJPEG_CALL(nvjpegDecodeJpegTransferToDevice(
+            handle_,
+            decoder,
+            state,
+            nvjpeg_state->jpeg_stream,
+            ws->stream()));
+
+        NVJPEG_CALL(nvjpegDecodeJpegDevice(
```

Comment:
Please add fallback to the CPU as well here.

Reply:
Done